### ISSUE TITLE
Scattergl does not always render markers

### ISSUE THREAD

**Issue by rhamner** [plotly/plotly.js#2999]
Occasionally when trying to render markers with scattergl, I get the following:

Offscreen-For-WebGL-000000C6A5E1D5E0]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 0

When this happens, most of the markers are not rendered. Hover and click still work for them though. An example of hovering over a point that doesn't render is included in the screenshot below:

![Image of fake hover](http://i65.tinypic.com/bfj9k6.png)

That plot looks like the following when all points are rendered:

![image](https://user-images.githubusercontent.com/20448486/45388752-85af6900-b5df-11e8-85ec-a9606079f42e.png)

If I switch to 'scatter' or 'pointcloud' or use lines instead of markers, it works fine (lines example):

![image](https://user-images.githubusercontent.com/20448486/45388678-497c0880-b5df-11e8-97de-81ffbe6c1cd0.png)

Stranger, I often get this error with a subset of the data but not with the full data set. The full data set looks like this using markers and scattergl:

![image](https://user-images.githubusercontent.com/20448486/45388724-6b758b00-b5df-11e8-9b4a-9a09609d97a2.png)

When I switch to 'edit in chart studio' the problem goes away. When I dump the plot data to json then bring into codepen, the problem goes away (https://codepen.io/rhamner/pen/zJWREL). It is probably important to know that this is the same machine...the issue appears with the same data set on the same machine sometimes.

Any idea on what could be causing this? Unfortunately, I can't get the issue to reproduce with the exact same data in codepen so I'm hoping the warning/error, screenshots, and actual data set in that codepen link are sufficient.

I saw a maybe similar issue reported here:

https://github.com/plotly/dash-core-components/issues/204

**Comment by etpinard** [comment]
What version of plotly.js are you using?

**Comment by rhamner** [comment]
1.40.1. I have tried using https://cdn.plot.ly/plotly-latest.min.js also.

**Comment by etpinard** [comment]
Ok, and are you using dash?

**Comment by rhamner** [comment]
I am not. Just straight javascript/plotly.js

**Comment by etpinard** [comment]
> I am not. Just straight javascript/plotly.js

Can you share a snippet of your work then to help us out?

**Comment by etpinard** [comment]
... or is the JS code in your https://codepen.io/rhamner/pen/zJWREL equivalent to the one that yield those 

`Offscreen-For-WebGL-000000C6A5E1D5E0]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 0` error messages?

**Comment by rhamner** [comment]
The code in the codepen is equivalent. That's the exact data and layout I'm using, and I'm using Plotly.react as the call.

**Comment by etpinard** [comment]
All right, well that one as is is going to be a little hard to debug.

Can you think of anything else (e.g in the html or css) that differs from your page and the codepen you posted?

**Comment by rhamner** [comment]
Ok...I thought of a way to reproduce it in codepen. It's order dependent...i.e., in our main app that shows it, it's when a single plot changes content that this occurs. I was able to mimic that in that codepen and here's the link again:

https://codepen.io/rhamner/pen/zJWREL

What should happen when you run this is that it plots a bunch of markers, then plots a subset as lines ~5 seconds later, then plots that same subset as markers after another ~5 seconds. That last marker plot should show the failure to render with the hover over non-rendered points.

**Comment by alexcjohnson** [comment]
Ha, I was about to write "crazy, I see the failure in the last plot in your codepen, one of the ones labeled 'Always seems to work'" but I guess that was as you were editing it. Small style point: we very much appreciate your effort to find a way to make this bug reproducible - thanks very much! but once you've posted a codepen here, it's better to fork it and edit the copy, so the written record here is clear on what we should expect to see at any given link.

I see the points disappear (but remain active for hover, as you describe) consistently on both Chrome and FF on my mac.

**Comment by rhamner** [comment]
Augh...I didn't think to fork. I'll fork and update the links after dinner. To confirm though...the link now shows the issue for you?

**Comment by alexcjohnson** [comment]
Yes, I see the issue repeatably now.

Don't worry about forking now - I almost didn't mention it because we really do appreciate your help in zeroing in on the problem. The state you left it in is great, it shows the bug in a clear and straightforward way. Just something to keep in mind for future issues 😄

**Comment by etpinard** [comment]
Thanks very much for your efforts @rhamner !

**Comment by kevinwuhoo** [comment]
I'm also seeing this. This seems to be a regression between 1.36.1 and 1.37.0 if that helps any. Changing the plotly.js version in the codepen seems to yield a working version. Thanks all!

**Comment by rhamner** [comment]
Hi guys...any updates on this one?

**Comment by rfried** [comment]
I'm also seeing this issue

**Comment by seansfkelley** [comment]
I'm seeing this sporadically on 1.39.2. It behaves like a race condition and is very hard to reproduce in my environment reliably (props to @rhamner for nailing down a repro).

**Comment by archmoj** [comment]
This [codepen](https://codepen.io/MojtabaSamimi/pen/vVroZV/?editors=1111) may be a better demonstration of this issue. This is to confirm that the problem only happens with `scattergl` not 'scatter' and while switching modes.

**Comment by rhamner** [comment]
Interestingly, that codepen does not show the issue for me @archmoj

**Comment by archmoj** [comment]
@rhamner Do you see 10 successful updates of the graph?

**Comment by slishak** [comment]
After a bit more investigation, this issue appears to be the root cause of https://github.com/plotly/dash-core-components/issues/313. When using plotly.js 1.40.0, the example Dash app in that issue works fine, but when upgrading to 1.40.1 the error is always triggered.

**Comment by etpinard** [comment]
Thanks @slishak , from @archmoj's investigation, we now think https://github.com/plotly/plotly.js/pull/2939 is the culprit.

Can I ask you (and everyone else here) which browser you are using?

**Comment by slishak** [comment]
I'm using Chrome on Windows 7.

**Comment by seansfkelley** [comment]
Chrome: Version 69.0.3497.100 (Official Build) (64-bit)
OS: macOS 10.14

**Comment by etpinard** [comment]
Most likely solved in https://github.com/plotly/plotly.js/pull/3132

Closing for now, but might have to be reopened if the problem persists in our upcoming `v1.42.0`

**Comment by archmoj** [comment]
@maksufa Could you please fork & modify this [codepen](https://codepen.io/MojtabaSamimi/pen/QJwRzW) or create another demo regarding the problem you noticed?

**Comment by slishak** [comment]
I think the issue that's unresolved is https://github.com/plotly/plotly.js/issues/2334, there are codepens there that still reproduce it with the latest version.

**Comment by archmoj** [comment]
Thanks @slishak for the info. I created [this new demo](https://codepen.io/MojtabaSamimi/pen/QJwRzW) for this other problem. What I observed "in the browsers' console" is that depending on the number of points & updating to/from various number of data points, one may notice either errors such as `Uncaught TypeError: Cannot read property 'sizeAvg' of undefined` or the WebGL warning mentioned above on Chrome and `TypeError: k.marker is undefined` on Firefox.
